ruler/app.py

```diff
@@ -0,0 +1,174 @@
+import json
+import logging
+import os
+
+import boto3
+from botocore.exceptions import ClientError
+from crhelper import CfnResource
+
+MISSING_PROPERTY_ERROR_MESSAGE = 'Parameter is required'
+MISSING_ENVIRONMENT_VARIABLE_MESSAGE = 'Environment variable is required'
+MISSING_CLIENT_SECRET_KEY = 'OIDC Client secret key not found in AWS System Manager parameter store'
+ALB_RULES_ACCESS_ERROR = 'Problem retrieving AWS application load balancer rules'
+ALB_RULE_CREATION_ERROR = 'Problem creating AWS application load balancer rule'
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+helper = CfnResource(
+  json_logging=False, log_level='DEBUG', boto_level='DEBUG')
+
+
+def get_variables(func, var_names, error_msg):
+  values = []
+
+  for var_name in var_names:
+    value = func(var_name)
+    if not value:
+      raise ValueError(f'{error_msg}: {var_name}')
+    values.append(value)
+
+  return values
+
+
+def get_properties(event):
+  resource_properties = event.get('ResourceProperties')
+  param_names = ['InstanceId', 'TargetGroupArn', 'ListenerArn']
+  variables = get_variables(resource_properties.get, param_names, MISSING_PROPERTY_ERROR_MESSAGE)
+  return variables
+
+
+def get_envvars():
+  env_var_names = [
+    'OIDC_CLIENT_SECRET_KEYNAME',
+    'OIDC_ISSUER',
+    'OIDC_AUTHORIZATION_ENDPOINT',
+    'OIDC_TOKEN_ENDPOINT',
+    'OIDC_USER_INFO_ENDPOINT',
+    'OIDC_CLIENT_ID'
+  ]
+  return get_variables(os.getenv, env_var_names, MISSING_ENVIRONMENT_VARIABLE_MESSAGE)
+
+
+def get_client(service):
+  return boto3.client(service)
+
+
+def get_client_key(key_name):
+  ssm = get_client('ssm')
+  try:
+    response = ssm.get_parameter(
+      Name=key_name,
+      WithDecryption=True)
+  except ClientError as e:
+    client_error_msg = e.response['Error']['Code']
+    exception_msg = f'{MISSING_CLIENT_SECRET_KEY}: key_name={key_name}; {client_error_msg}'
+    raise Exception(exception_msg)
+
+  return response['Parameter']['Value']
+
+
+def next_priority(listener_arn):
+  elbv2 = get_client('elbv2')
+  try:
+    rules = elbv2.describe_rules(ListenerArn=listener_arn).get('Rules')
+  except ClientError as e:
+    client_error_msg = e.response['Error']['Code']
+    exception_msg = f'{ALB_RULES_ACCESS_ERROR}: listener_arn={listener_arn}; {client_error_msg}'
+    raise Exception(exception_msg)
+
+  priorities = sorted([int(rule['Priority']) for rule in rules if rule['Priority'] != 'default'])
+  priority = 1 if not priorities else priorities[-1] + 1
```

Comment:
could we get into a situation where we may have a non-contiguous priorities list?  for example in the process of adding and removing rules you end up with rules having priorities `1,2,10,30,70,300,1000`?  This would make your next priority value 1001?  

Also I'm wondering what significance the priority has for our use case?  Does it matter whether an instance is set to priority 100 vs 1000?

Reply:
I believe we tested this and found that the priorities adjust when one is deleted from the middle.

Reply:
1. During tests when a lower priority was removed the others were adjusted. That is why I included no logic to sort through all the priorities and start resetting them, or finding the first gap.
2. Currently in our case the priority has no significance, but it is required.

Reply:
AWS automatically adjusting priorities, especially without you knowing itm does not make sense to me.  I mean if I setup a bunch of rules in an ALB with explicit priority levels then I would expect them to stay the same.   I would not be happy AWS just re-adjusted them without me know it.  What do you think @xschildw?

Reply:
Well, that is how it works. I encourage you to go make a rule, then make another with the next priority number, remove the first, and see what happens. Please note Khai, this is emphatically desirable behavior for our purpose, otherwise the logic to adjust priorities would have to be done on clients, and then imagine multiple clients trying to do so, and imagine the cluster event that would be.